Pre-commit hooks don't work on python 3.8

The version of `typed_ast` pinned in the `pre-commit` config doesn't install on python 3.8 due to `pgenheaders.h` being removed from python as of `3.8a4` (see [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1709774) for more).

<details>
```console
git commit -S -s                                                               714ms  Tue 03 Mar 2020 10:36:21 AM EST
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/techalchemy/.cache/pre-commit/patch1583249782.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/ambv/black.
[INFO] Initializing environment for https://github.com/asottile/pyupgrade.
[INFO] Initializing environment for https://github.com/asottile/seed-isort-config.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-isort.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/ambv/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/pyupgrade.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/seed-isort-config.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-mypy.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Restored changes from /home/techalchemy/.cache/pre-commit/patch1583249782.
An unexpected error has occurred: CalledProcessError: command: ('/home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/bin/python', '/home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/bin/pip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing /home/techalchemy/.cache/pre-commit/repo4463pk7w
    Collecting mypy==0.610
      Using cached mypy-0.610-py3-none-any.whl (1.3 MB)
    Collecting typed-ast<1.2.0,>=1.1.0
      Using cached typed-ast-1.1.2.tar.gz (202 kB)
    Building wheels for collected packages: pre-commit-dummy-package, typed-ast
      Building wheel for pre-commit-dummy-package (setup.py): started
      Building wheel for pre-commit-dummy-package (setup.py): finished with status 'done'
      Created wheel for pre-commit-dummy-package: filename=pre_commit_dummy_package-0.0.0-py3-none-any.whl size=1178 sha256=d98d8c1e466160a9883ad37731a032c28e9a7d3008ed735073e735e482427cfa
      Stored in directory: /tmp/pip-ephem-wheel-cache-2cks3hh3/wheels/19/11/e0/2b4589774cd6cf26b3fca589b7798d2ad074639b3653278790
      Building wheel for typed-ast (setup.py): started
      Building wheel for typed-ast (setup.py): finished with status 'error'
      Running setup.py clean for typed-ast
    Successfully built pre-commit-dummy-package
    Failed to build typed-ast
    Installing collected packages: typed-ast, mypy, pre-commit-dummy-package
        Running setup.py install for typed-ast: started
        Running setup.py install for typed-ast: finished with status 'error'
    
stderr:
      ERROR: Command errored out with exit status 1:
       command: /home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"'; __file__='"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-_w72jhe7
           cwd: /tmp/pip-install-wzqnea69/typed-ast/
      Complete output (25 lines):
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.8
      creating build/lib.linux-x86_64-3.8/typed_ast
      copying typed_ast/__init__.py -> build/lib.linux-x86_64-3.8/typed_ast
      copying typed_ast/ast27.py -> build/lib.linux-x86_64-3.8/typed_ast
      copying typed_ast/conversions.py -> build/lib.linux-x86_64-3.8/typed_ast
      copying typed_ast/ast3.py -> build/lib.linux-x86_64-3.8/typed_ast
      warning: build_py: byte-compiling is disabled, skipping.
      
      running build_ext
      building '_ast27' extension
      creating build/temp.linux-x86_64-3.8
      creating build/temp.linux-x86_64-3.8/ast27
      creating build/temp.linux-x86_64-3.8/ast27/Parser
      creating build/temp.linux-x86_64-3.8/ast27/Python
      creating build/temp.linux-x86_64-3.8/ast27/Custom
      gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Iast27/Include -I/home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/include -I/home/techalchemy/.pyenv/versions/3.8.2/include/python3.8 -c ast27/Parser/acceler.c -o build/temp.linux-x86_64-3.8/ast27/Parser/acceler.o
      ast27/Parser/acceler.c:13:10: fatal error: pgenheaders.h: No such file or directory
         13 | #include "pgenheaders.h"
            |          ^~~~~~~~~~~~~~~
      compilation terminated.
      error: command 'gcc' failed with exit status 1
      ----------------------------------------
      ERROR: Failed building wheel for typed-ast
        ERROR: Command errored out with exit status 1:
         command: /home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"'; __file__='"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-hz_trups/install-record.txt --single-version-externally-managed --compile --install-headers /home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/include/site/python3.8/typed-ast
             cwd: /tmp/pip-install-wzqnea69/typed-ast/
        Complete output (25 lines):
        running install
        running build
        running build_py
        creating build
        creating build/lib.linux-x86_64-3.8
        creating build/lib.linux-x86_64-3.8/typed_ast
        copying typed_ast/__init__.py -> build/lib.linux-x86_64-3.8/typed_ast
        copying typed_ast/ast27.py -> build/lib.linux-x86_64-3.8/typed_ast
        copying typed_ast/conversions.py -> build/lib.linux-x86_64-3.8/typed_ast
        copying typed_ast/ast3.py -> build/lib.linux-x86_64-3.8/typed_ast
        warning: build_py: byte-compiling is disabled, skipping.
        
        running build_ext
        building '_ast27' extension
        creating build/temp.linux-x86_64-3.8
        creating build/temp.linux-x86_64-3.8/ast27
        creating build/temp.linux-x86_64-3.8/ast27/Parser
        creating build/temp.linux-x86_64-3.8/ast27/Python
        creating build/temp.linux-x86_64-3.8/ast27/Custom
        gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -Iast27/Include -I/home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/include -I/home/techalchemy/.pyenv/versions/3.8.2/include/python3.8 -c ast27/Parser/acceler.c -o build/temp.linux-x86_64-3.8/ast27/Parser/acceler.o
        ast27/Parser/acceler.c:13:10: fatal error: pgenheaders.h: No such file or directory
           13 | #include "pgenheaders.h"
              |          ^~~~~~~~~~~~~~~
        compilation terminated.
        error: command 'gcc' failed with exit status 1
        ----------------------------------------
    ERROR: Command errored out with exit status 1: /home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"'; __file__='"'"'/tmp/pip-install-wzqnea69/typed-ast/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-hz_trups/install-record.txt --single-version-externally-managed --compile --install-headers /home/techalchemy/.cache/pre-commit/repo4463pk7w/py_env-python3.8/include/site/python3.8/typed-ast Check the logs for full command output.
    
Check the log at /home/techalchemy/.cache/pre-commit/pre-commit.log
```
</details>

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,6 +1,6 @@
 repos:
 -   repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v1.3.0
+    rev: v2.5.0
     hooks:
     -   id: trailing-whitespace
     -   id: end-of-file-fixer
@@ -8,26 +8,26 @@ repos:
     -   id: check-added-large-files
         exclude: ^docs/conf.py$
 -   repo: https://github.com/ambv/black
-    rev: 18.9b0
+    rev: 19.10b0
     hooks:
     -   id: black
         args: [--py36]
 -   repo: https://github.com/asottile/pyupgrade
-    rev: v1.3.1
+    rev: v2.1.0
     hooks:
     -   id: pyupgrade
         args: [--py36-plus]
 -   repo: https://github.com/asottile/seed-isort-config
-    rev: v1.0.1
+    rev: v2.1.0
     hooks:
     -   id: seed-isort-config
         args: [--application-directories, '.:src']
 -   repo: https://github.com/pre-commit/mirrors-isort
-    rev: v4.3.4
+    rev: v4.3.21
     hooks:
     -   id: isort
 -   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.610-1
+    rev: v0.761
     hooks:
     -   id: mypy
         exclude: ^docs/conf.py$

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -11,7 +11,7 @@ repos:
     rev: 19.10b0
     hooks:
     -   id: black
-        args: [--py36]
+        args: [--target-version, py36]
 -   repo: https://github.com/asottile/pyupgrade
     rev: v2.1.0
     hooks:

--- a/docs/conf.py
+++ b/docs/conf.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 import os
 
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -83,3 +83,14 @@ doc_build =
     recommonmark
     # git+https://github.com/pypa/pypa-docs-theme.git#egg=pypa-docs-theme
     # git+https://github.com/python/python-docs-theme.git#egg=python-docs-theme
+
+[isort]
+atomic = true
+not_skip = __init__.py
+line_length = 88
+multi_line_output = 3
+known_third_party = aiohttp,aiohttp_xmlrpc,asynctest,filelock,freezegun,packaging,pkg_resources,pytest,setuptools
+known_first_party = bandersnatch,bandersnatch_filter_plugins
+force_grid_wrap = 0
+use_parentheses=True
+include_trailing_comma = True

--- a/src/bandersnatch/configuration.py
+++ b/src/bandersnatch/configuration.py
@@ -24,7 +24,7 @@ class Singleton(type):  # pragma: no cover
 
     def __call__(cls, *args: Any, **kwargs: Any) -> Type:
         if cls not in cls._instances:
-            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+            cls._instances[cls] = super().__call__(*args, **kwargs)
         return cls._instances[cls]
 
 

--- a/src/bandersnatch/main.py
+++ b/src/bandersnatch/main.py
@@ -160,7 +160,7 @@ def main(loop: Optional[asyncio.AbstractEventLoop] = None) -> int:
         default_config_path = Path(__file__).parent / "default.conf"
         try:
             shutil.copy(default_config_path, args.config)
-        except IOError as e:
+        except OSError as e:
             logger.error(f"Could not create config file: {e}")
         return 1
 

--- a/src/bandersnatch/mirror.py
+++ b/src/bandersnatch/mirror.py
@@ -369,7 +369,7 @@ class Mirror:
             logger.info("Generation file inconsistent. Reinitialising status files.")
             self._reset_mirror_status()
             generation = CURRENT_GENERATION
-        except IOError:
+        except OSError:
             logger.info("Generation file missing. Reinitialising status files.")
             # This is basically the 'install' generation: anything previous to
             # release 1.0.2.
@@ -483,7 +483,7 @@ async def mirror(config: configparser.ConfigParser) -> int:  # noqa: C901
         # This works around "TypeError: object
         # MagicMock can't be used in 'await' expression"
         changed_packages: Dict[str, Set[str]] = {}
-        if not isinstance(mirror, Mock):
+        if not isinstance(mirror, Mock):  # type: ignore
             changed_packages = await mirror.synchronize()
         logger.info(f"{len(changed_packages)} packages had changes")
         for package_name, changes in changed_packages.items():

--- a/src/bandersnatch/tests/test_mirror.py
+++ b/src/bandersnatch/tests/test_mirror.py
@@ -86,7 +86,7 @@ def test_mirror_generation_3_resets_status_files(tmpdir):
     assert m.synced_serial == 0
     assert not os.path.exists(str(tmpdir / "todo"))
     assert not os.path.exists(str(tmpdir / "status"))
-    assert open(str(tmpdir / "generation"), "r").read() == "5"
+    assert open(str(tmpdir / "generation")).read() == "5"
 
 
 def test_mirror_generation_4_resets_status_files(tmpdir):
@@ -101,7 +101,7 @@ def test_mirror_generation_4_resets_status_files(tmpdir):
     assert m.synced_serial == 0
     assert not os.path.exists(str(tmpdir / "todo"))
     assert not os.path.exists(str(tmpdir / "status"))
-    assert open(str(tmpdir / "generation"), "r").read() == "5"
+    assert open(str(tmpdir / "generation")).read() == "5"
 
 
 def test_mirror_filter_packages_match(tmpdir):

--- a/src/bandersnatch/tests/test_package.py
+++ b/src/bandersnatch/tests/test_package.py
@@ -469,7 +469,7 @@ async def test_survives_exceptions_from_record_finished_package(mirror):
     def record_finished_package(name):
         import errno
 
-        raise IOError(errno.EBADF, "Some transient error?")
+        raise OSError(errno.EBADF, "Some transient error?")
 
     mirror.packages_to_sync = {"Foo": 1}
     mirror.record_finished_package = record_finished_package

--- a/src/bandersnatch/tests/test_utils.py
+++ b/src/bandersnatch/tests/test_utils.py
@@ -82,7 +82,7 @@ def test_rewrite_nonexisting_file(tmpdir, monkeypatch):
     monkeypatch.chdir(tmpdir)
     with rewrite("sample", "w") as f:
         f.write("csdf")
-    with open("sample", "r") as f:
+    with open("sample") as f:
         assert f.read() == "csdf"
 
 

--- a/tox.ini
+++ b/tox.ini
@@ -35,4 +35,4 @@ commands=
 
 [isort]
 include_trailing_comma = True
-known_third_party = aiohttp,aiohttp_xmlrpc,asynctest,cloud_sptheme,filelock,freezegun,guzzle_sphinx_theme,packaging,pkg_resources,pypa_theme,pytest,recommonmark,setuptools,sphinx_bootstrap_theme,sphinx_rtd_theme
+known_third_party = aiohttp,aiohttp_xmlrpc,asynctest,filelock,freezegun,packaging,pkg_resources,pytest,setuptools

--- a/tox.ini
+++ b/tox.ini
@@ -34,5 +34,12 @@ commands=
     flake8
 
 [isort]
-include_trailing_comma = True
+atomic = true
+not_skip = __init__.py
+line_length = 88
+multi_line_output = 3
 known_third_party = aiohttp,aiohttp_xmlrpc,asynctest,filelock,freezegun,packaging,pkg_resources,pytest,setuptools
+known_first_party = bandersnatch,bandersnatch_filter_plugins
+force_grid_wrap = 0
+use_parentheses=True
+include_trailing_comma = True

